File: solutions/clojure/120-lowest-common-ancestor-of-a-binary-search-tree.clj
;; Problem 120: Lowest Common Ancestor of a Binary Search Tree (LeetCode 235)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn lca-bst [root p q]
  (cond
    (nil? root) nil
    (and (< p (:val root)) (< q (:val root)))
      (lca-bst (:left root) p q)
    (and (> p (:val root)) (> q (:val root)))
      (lca-bst (:right root) p q)
    :else (:val root)))

File: solutions/typescript/120-lowest-common-ancestor-of-a-binary-search-tree.ts
/**
 * Problem 120: Lowest Common Ancestor of a Binary Search Tree (LeetCode 235)
 * Difficulty: Med
 * Language: TypeScript
 */
interface TreeNode { val: number; left: TreeNode | null; right: TreeNode | null; }
function lowestCommonAncestorBST(root: TreeNode | null, p: number, q: number): number | null {
  if (root === null) return null;
  if (p < root.val && q < root.val) return lowestCommonAncestorBST(root.left, p, q);
  if (p > root.val && q > root.val) return lowestCommonAncestorBST(root.right, p, q);
  return root.val;
}
